### PR TITLE
Fix float comparison precision

### PR DESCRIPTION
**Pull request by mvorisek**
needed for https://github.com/sebastianbergmann/phpunit/pull/4972

fixed EPSILON was problematic and small values like `1E-100` and `1E-200` were considered the same

this PR removes the `DoubleComparator` completely, if number is supposed to be compared unexactly, delta can be passed to `NumericComparator`

### CODE REVIEW

File: src/DoubleComparator.php [hunk 1]
```diff
@@ -1,56 +0,0 @@
...
- *
- * For the full copyright and license information, please view the LICENSE
- * file that was distributed with this source code.
- */
-namespace SebastianBergmann\Comparator;
-
-/**
- * Compares doubles for equality.
- */
-class DoubleComparator extends NumericComparator
```
Comment (sebastianbergmann) [inline src/DoubleComparator.php:15]: The removal of `DoubleComparator` is a BC break. We should keep the class around and just do not use it anymore.

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,4 @@
 /.php_cs.cache
 /composer.lock
 /vendor
+/.phpunit.result.cache
```
[the remark this answers is not in the document]
Comment (mvorisek) [reply .gitignore]: done

File: src/DoubleComparator.php (hunk 1 above)
Comment (mvorisek) [reply src/DoubleComparator.php:15]: done and marked as deprecated, when merging to `main`, `DoubleComparator` and `NumericComparatorTest` classes/files should be deleted

**Comment by sebastianbergmann** [comment]
The following test failures in PHPUnit 8.5's test suite now occur:

```
1) PHPUnit\Framework\AssertTest::testAssertEqualsSucceeds with data set #19 ('0.33333333333333', 0.33333333333333337)
Failed asserting that 0.33333333333333337 matches expected '0.33333333333333'.

/usr/local/src/phpunit/src/Framework/Constraint/IsEqual.php:100
/usr/local/src/phpunit/src/Framework/Assert.php:2952
/usr/local/src/phpunit/src/Framework/Assert.php:635
/usr/local/src/phpunit/tests/unit/Framework/AssertTest.php:266
/usr/local/src/phpunit/src/Framework/TestCase.php:1492
/usr/local/src/phpunit/src/Framework/TestCase.php:1112
/usr/local/src/phpunit/src/Framework/TestResult.php:703
/usr/local/src/phpunit/src/Framework/TestCase.php:838
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/TextUI/TestRunner.php:659
/usr/local/src/phpunit/src/TextUI/Command.php:235
/usr/local/src/phpunit/src/TextUI/Command.php:194

2) PHPUnit\Framework\AssertTest::testAssertEqualsSucceeds with data set #20 (0.3333333333333333, '0.33333333333333')
Failed asserting that '0.33333333333333' matches expected 0.3333333333333333.

/usr/local/src/phpunit/src/Framework/Constraint/IsEqual.php:100
/usr/local/src/phpunit/src/Framework/Assert.php:2952
/usr/local/src/phpunit/src/Framework/Assert.php:635
/usr/local/src/phpunit/tests/unit/Framework/AssertTest.php:266
/usr/local/src/phpunit/src/Framework/TestCase.php:1492
/usr/local/src/phpunit/src/Framework/TestCase.php:1112
/usr/local/src/phpunit/src/Framework/TestResult.php:703
/usr/local/src/phpunit/src/Framework/TestCase.php:838
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/TextUI/TestRunner.php:659
/usr/local/src/phpunit/src/TextUI/Command.php:235
/usr/local/src/phpunit/src/TextUI/Command.php:194

3) PHPUnit\Framework\AssertTest::testAssertEqualsSucceeds with data set #27 (0.3333333333333333, 0.33333333333333337)
Failed asserting that 0.33333333333333337 matches expected 0.3333333333333333.

/usr/local/src/phpunit/src/Framework/Constraint/IsEqual.php:100
/usr/local/src/phpunit/src/Framework/Assert.php:2952
/usr/local/src/phpunit/src/Framework/Assert.php:635
/usr/local/src/phpunit/tests/unit/Framework/AssertTest.php:266
/usr/local/src/phpunit/src/Framework/TestCase.php:1492
/usr/local/src/phpunit/src/Framework/TestCase.php:1112
/usr/local/src/phpunit/src/Framework/TestResult.php:703
/usr/local/src/phpunit/src/Framework/TestCase.php:838
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/TextUI/TestRunner.php:659
/usr/local/src/phpunit/src/TextUI/Command.php:235
/usr/local/src/phpunit/src/TextUI/Command.php:194

4) PHPUnit\Framework\AssertTest::testAssertNotEqualsFails with data set #19 ('0.33333333333333', 0.33333333333333337)
Failed asserting that exception of type "PHPUnit\Framework\AssertionFailedError" is thrown.

/usr/local/src/phpunit/src/Framework/Constraint/Constraint.php:120
/usr/local/src/phpunit/src/Framework/Constraint/Constraint.php:55
/usr/local/src/phpunit/src/Framework/Assert.php:2952
/usr/local/src/phpunit/src/Framework/TestCase.php:1555
/usr/local/src/phpunit/src/Framework/TestCase.php:1112
/usr/local/src/phpunit/src/Framework/TestResult.php:703
/usr/local/src/phpunit/src/Framework/TestCase.php:838
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/TextUI/TestRunner.php:659
/usr/local/src/phpunit/src/TextUI/Command.php:235
/usr/local/src/phpunit/src/TextUI/Command.php:194

5) PHPUnit\Framework\AssertTest::testAssertNotEqualsFails with data set #20 (0.3333333333333333, '0.33333333333333')
Failed asserting that exception of type "PHPUnit\Framework\AssertionFailedError" is thrown.

/usr/local/src/phpunit/src/Framework/Constraint/Constraint.php:120
/usr/local/src/phpunit/src/Framework/Constraint/Constraint.php:55
/usr/local/src/phpunit/src/Framework/Assert.php:2952
/usr/local/src/phpunit/src/Framework/TestCase.php:1555
/usr/local/src/phpunit/src/Framework/TestCase.php:1112
/usr/local/src/phpunit/src/Framework/TestResult.php:703
/usr/local/src/phpunit/src/Framework/TestCase.php:838
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/TextUI/TestRunner.php:659
/usr/local/src/phpunit/src/TextUI/Command.php:235
/usr/local/src/phpunit/src/TextUI/Command.php:194

6) PHPUnit\Framework\AssertTest::testAssertNotEqualsFails with data set #27 (0.3333333333333333, 0.33333333333333337)
Failed asserting that exception of type "PHPUnit\Framework\AssertionFailedError" is thrown.

/usr/local/src/phpunit/src/Framework/Constraint/Constraint.php:120
/usr/local/src/phpunit/src/Framework/Constraint/Constraint.php:55
/usr/local/src/phpunit/src/Framework/Assert.php:2952
/usr/local/src/phpunit/src/Framework/TestCase.php:1555
/usr/local/src/phpunit/src/Framework/TestCase.php:1112
/usr/local/src/phpunit/src/Framework/TestResult.php:703
/usr/local/src/phpunit/src/Framework/TestCase.php:838
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/Framework/TestSuite.php:619
/usr/local/src/phpunit/src/TextUI/TestRunner.php:659
/usr/local/src/phpunit/src/TextUI/Command.php:235
/usr/local/src/phpunit/src/TextUI/Command.php:194
```

@mvorisek Can you please send a PR to PHPUnit's `8.5` branch to address this? Thanks!

**Comment by olivernybroe** [comment]
This change actually broke multiple of our tests in a minor version. 

Wasn't that big a problem to fix, but took quite a while to troubleshoot why our tests suddenly gave errors like these
image

**Comment by mvorisek** [comment]
> This change actually broke multiple of our tests in a minor version.
> 
> Wasn't that big a problem to fix, but took quite a while to troubleshoot why our tests suddenly gave errors like these image

this PR is continuation of https://github.com/sebastianbergmann/phpunit/commit/136fae39306c8e6cda3fb5ec8fcc0ff8cff8cf4e which effectively implies zero delta for numbers greater than 1 - that is why you hit this problem

also, casting from float to string depends on php.ini `precision`, in phpunit, this problem is fixed now and the full precision is always displayed - in the output you posted, the `Actual: 0.1505` should display the full float precision, please submit an issue to the phpunit printer console you use - see exporter https://github.com/sebastianbergmann/exporter/commit/a9f4590ffb748ac24beffc5d18cda991f43f0285 fix

**Comment by olivernybroe** [comment]
@mvorisek Yep wasn't much of work to fix the issues when it became clear why 👍 

Also big fan of this change, just wanted to add a comment here for people in the future hitting this.

**Pull request merged**